File: igusto/igusto/page/room_order/room_order.py
import frappe
import json
from frappe.utils import now_datetime, nowdate

@frappe.whitelist()
def get_menu_items():
    """Fetch active items from Item doctype."""
    items = frappe.get_all(
        "Item",
        fields=["name", "item_name"],
        filters={"disabled": 0},
        order_by="item_name asc"
    )
    return items


@frappe.whitelist()
def get_service_items():
    """Return static room service options."""
    return ["Towel Change", "Cleaning", "Extra Pillow"]


@frappe.whitelist()
def get_spa_items():
    """Static spa service options (checkbox style)."""
    return ["Body Massage", "Head Massage", "Foot Spa", "Sauna", "Facial Therapy"]


@frappe.whitelist()
def get_laundry_items():
    return ["Clothes Wash", "Ironing", "Dry Clean", "Fold & Pack"]


@frappe.whitelist()
def get_transport_items():
    return ["Airport Pickup", "Airport Drop", "City Tour", "Cab on Call"]


@frappe.whitelist()
def get_room_numbers():
    """
    Return a list of room_number values from Guest Onboarding doctype.
    Assumes Guest Onboarding doctype has field `room_number`.
    """
    try:
        rows = frappe.get_all("Guest Onboarding", fields=["room_number"], filters={"room_number": ["!=", ""]}, order_by="room_number asc")
        return [r.room_number for r in rows]
    except Exception:
        # fallback: return empty list if doctype not present
        return []


@frappe.whitelist()
def get_guest_by_room(room_number):
    """
    Return guest (name) for given room_number from Guest Onboarding.
    Assumes Guest Onboarding has `guest` field which may be Guest ID or guest name.
    """
    if not room_number:
        return None
    guest = frappe.db.get_value("Guest Onboarding", {"room_number": room_number}, "guest")
    # if stored guest is a Link to Guest doctype, try to fetch full_name
    if guest and frappe.db.exists("Guest", guest):
        full_name = frappe.db.get_value("Guest", guest, "full_name")
        return full_name or guest
    return guest


@frappe.whitelist()
def get_item_rate(item_name):
    """
    Return a preview rate for an item_name (by Item.item_name lookup).
    Falls back to 0 if not found.
    """
    if not item_name:
        return 0
    item_code = frappe.db.get_value("Item", {"item_name": item_name}, "name")
    if item_code:
        # common field in custom setups: standard_rate
        rate = frappe.db.get_value("Item", item_code, "standard_rate")
        if rate is None:
            # try other common fields
            rate = frappe.db.get_value("Item", item_code, "last_purchase_rate")
        return rate or 0
    return 0

@frappe.whitelist(allow_guest=True)
def get_company_details():
    company = frappe.get_all(
        "Company",
        fields=["name", "company_description"],
        limit=1
    )
    if not company:
        return {}

    company = company[0]

    # Always use company_description as address
    address_text = (company.get("company_description") or "").strip()

    return {
        "company_name": company.name,  
        "address": address_text, 
        "logo": _get_company_logo(company.name)
    }

def _get_company_logo(company_name):
    file = frappe.get_all(
        "File",
        filters={
            "attached_to_doctype": "Company",
            "attached_to_name": company_name,
            "is_private": 0
        },
        fields=["file_url"],
        limit=1
    )
    return file[0].file_url if file else ""

@frappe.whitelist()
def create_room_order(data):
    """
    Create a new Room Order or Task (for Room Service) document with child table entries.
    Auto-sets current date and guest name as room_order_customer.
    """
    try:
        data = json.loads(data)
        service_type = data.get("service_type")

        # Helper: get item_code
        def get_item_code(item_name):
            if not item_name:
                return None
            return frappe.db.get_value("Item", {"item_name": item_name}, "name")

        # Helper: ensure Task Type exists
        def ensure_task_type(task_type_name):
            """Safely create Task Type if it doesn't exist."""
            if not task_type_name:
                return

            # If already exists, no need to create
            if frappe.db.exists("Task Type", task_type_name):
                return

            # Create a new Task Type doc
            tt = frappe.new_doc("Task Type")

            # Try all naming possibilities (covers all ERPNext setups)
            try:
                tt.task_type = task_type_name
            except Exception:
                pass

            try:
                tt.name = task_type_name
            except Exception:
                pass

            try:
                tt.__newname = task_type_name
            except Exception:
                pass

            # Fallback if none of the above fields work
            if not tt.name:
                tt.set("name", task_type_name)

            tt.insert(ignore_permissions=True)

        # ------------------ ROOM SERVICE FLOW ------------------
        if service_type == "Room Service":
            items = data.get("items", [])
            guest = data.get("guest")
            room_no = data.get("room_number")
            delivery_to = data.get("delivery_to")

            created_tasks = []

            # --- Create separate Task for each item ---
            for i in items:
                item_name = i.get("item_name")
                if not item_name:
                    continue

                # Ensure Task Type exists
                ensure_task_type(item_name)

                # Create Task
                task = frappe.new_doc("Task")
                task.subject = f"{item_name} - Room {room_no}"
                task.status = "Open"
                task.task_type = item_name  # ✅ show as Task Type in UI
                task.room_number = room_no
                task.delivery_to = delivery_to

                # Description with details
                task.description = (
                    f"<b>Guest:</b> {guest}<br>"
                    f"<b>Room:</b> {room_no}<br>"
                    f"<b>Requested Service:</b> {service_type}<br>"
                    f"<b>Item:</b> {item_name}<br>"
                )

                task.insert(ignore_permissions=True)
                created_tasks.append(task.name)

            frappe.db.commit()

            # --- Create ONE Sales Order for all items ---
            so = frappe.new_doc("Sales Order")
            so.customer = guest
            so.delivery_date = nowdate()

            for i in items:
                item_code = get_item_code(i.get("item_name"))
                if not item_code:
                    continue
                so.append("items", {
                    "item_code": item_code,
                    "item_name": i.get("item_name"),
                    "qty": i.get("quantity") or i.get("qty") or 1,
                    "rate": i.get("rate") or 0,
                    "description": f"{service_type} - {i.get('custom_remarks') or ''}"
                })

            so.insert(ignore_permissions=True)
            frappe.db.commit()

            return {
                "status": "success",
                "message": f"{len(created_tasks)} Tasks and Sales Order {so.name} created successfully.",
                "tasks": created_tasks,
                "sales_order": so.name
            }

        # ------------------ OTHER SERVICE TYPES ------------------
        room_order = frappe.new_doc("Room Orders")
        room_order.room_number = data.get("room_number")
        room_order.room_order_date = nowdate()
        room_order.room_order_customer = data.get("guest")
        room_order.guest = data.get("guest")
        room_order.delivery_to = data.get("delivery_to")
        room_order.service_type = service_type

        items = data.get("items", [])
        for i in items:
            qty_value = i.get("passengers") or i.get("qty") or 1
            room_order.append("room_order_booking_items", {
                "item": i.get("item_name"),
                "qty": qty_value,
                "rate": i.get("rate") or 0,
                "custom_remarks": i.get("custom_remarks") or "",
                "category": service_type or "General"
            })

        room_order.insert(ignore_permissions=True)
        frappe.db.commit()

        # --- Create Sales Order linked to this Room Order ---
        so = frappe.new_doc("Sales Order")
        so.customer = data.get("guest")
        so.delivery_date = nowdate()

        for i in items:
            item_code = get_item_code(i.get("item_name"))
            if not item_code:
                continue

            so.append("items", {
                "item_code": item_code,
                "item_name": i.get("item_name"),
                "qty": i.get("quantity") or i.get("qty") or 1,
                "rate": i.get("rate") or 0,
                "description": f"{service_type} - {i.get('custom_remarks') or ''}"
            })

        so.insert(ignore_permissions=True)
        frappe.db.commit()

        return {
            "status": "success",
            "message": f"Room Order {room_order.name} and Sales Order {so.name} created successfully.",
            "name": room_order.name,
            "sales_order": so.name
        }

    except Exception as e:
        frappe.log_error(frappe.get_traceback(), "Create Room Order Error")
        frappe.throw(f"Error creating Room Order: {str(e)}")
        return {
            "status": "error",
            "message": str(e)
        }

File: igusto/igusto/page/room_order/room_order.js
frappe.pages['room-order'].on_page_load = function (wrapper) {
  new RoomOrder(wrapper);
};

class RoomOrder {
  constructor(wrapper) {
    this.page = frappe.ui.make_app_page({
      parent: wrapper,
      title: '',
      single_column: true
    });
    this.make();
    this.load_company_details();
    this.load_room_numbers();

  }
  load_company_details() {
    frappe.call({
      method: "igusto.igusto.page.room_order.room_order.get_company_details",
      callback: (r) => {
        const data = r.message;
        if (!data) return;

        const logo_html = data.logo
          ? `<img src="${data.logo}" class="company-logo">`
          : `<div class="company-logo-placeholder">No Logo</div>`;

        let address_html = data.address ? data.address : "";

const hasCustomAddress = data.custom_address && data.custom_address.trim() !== "";

let contact_html = "";

// Show phone + email only when custom_address is NOT present
if (!hasCustomAddress) {
  if (typeof data.phone === "string" && data.phone.trim() !== "") {
    contact_html += data.phone;
  }

  if (typeof data.email === "string" && data.email.trim() !== "") {
    // Add comma only if phone also exists
    contact_html += (contact_html ? ", " : "") + data.email;
  }
}

        const header_html = `
  <div class="company-header-inner">
    <div class="company-left">${logo_html}</div>
    <div class="company-right">
      <h2 class="company-name">${data.company_name}</h2>
      <div class="company-details">
        ${address_html ? `<div>${address_html}</div>` : ""}
        ${contact_html ? `<div>${contact_html}</div>` : ""}
      </div>
    </div>
  </div>
`;

        $(".company-header-wrapper").remove();
        $(".combined-card .company-header").html(header_html);

      }
    });
  }


  make() {
    $(frappe.render_template("room_order", {})).appendTo(this.page.body);

    const urlParams = new URLSearchParams(window.location.search);
    const qr_room_number = urlParams.get("room_number");

    if (qr_room_number) {
      // Fill "Order From" field
      const $orderFrom = $("input[placeholder='Enter Room Number']");
      $orderFrom
        .val(qr_room_number)
        .prop("readonly", true)
        .css({
          background: "#f5f5f5",
          fontWeight: "600",
          color: "#333",
        });

      // frappe.show_alert({
      //   message: `Room ${qr_room_number} detected from QR!`,
      //   indicator: "green",
      // });

      // Fetch and fill guest name
      frappe.call({
        method: "igusto.igusto.page.room_order.room_order.get_guest_by_room",
        args: { room_number: qr_room_number },
        callback: function (r) {
          console.log("Guest Response:", r);
          if (r.message) {
            $("#guest_name").val(r.message);
          } else {
            frappe.msgprint("No guest found for this room.");
          }
        }
      });
    }

    // listeners
    $("#service_type").on("change", (e) => {
      this.render_dynamic_fields(e.target.value);
    });

    $("#room_number").on("change", (e) => {
      const room = e.target.value;
      if (room) {
        this.fetch_guest_for_room(room);
      } else {
        $("#guest_name").val("");
      }
    });

    $("#submit_order").on("click", () => {
      this.submit_order();
    });
  }

  load_room_numbers() {
    frappe.call({
      method: "igusto.igusto.page.room_order.room_order.get_room_numbers",
      callback: (r) => {
        if (r.message) {
          const sel = $("#room_number");
          sel.empty();
          sel.append(`<option value="">Select Room Number</option>`);
          r.message.forEach(rn => {
            sel.append(`<option value="${frappe.utils.escape_html(rn)}">${frappe.utils.escape_html(rn)}</option>`);
          });
        }
      }
    });
  }

  fetch_guest_for_room(room_number) {
    frappe.call({
      method: "igusto.igusto.page.room_order.room_order.get_guest_by_room",
      args: { room_number },
      callback: (r) => {
        if (r.message) {
          $("#guest_name").val(r.message);
        } else {
          $("#guest_name").val("");
          frappe.msgprint("No guest found for selected room.");
        }
      }
    });
  }

  render_dynamic_fields(service_type) {
    const container = $("#dynamic_input");
    container.empty();

    if (!service_type) return;

    // RESTAURANT
    if (service_type === "Restaurant") {
      frappe.call({
        method: "igusto.igusto.page.room_order.room_order.get_menu_items",
        callback: (r) => {
          if (r.message) {
            let html = `
        <div class="child-table-wrapper">
          <table class="child-table" id="restaurant_table">
            <thead>
              <tr>
                <th>Item</th>
                <th>Rate</th>
                <th>Qty</th>
                <th>Amount</th>
                <th>Remarks</th>
                <th></th>
              </tr>
            </thead>
            <tbody></tbody>
          </table>
        </div>

        <button type="button" class="add-row-btn" id="add_row_btn">+ Add Row</button>
      `;

            container.html(html);

            const addRow = () => {
              const options = r.message
                .map(i => `<option value="${i.item_name}">${i.item_name}</option>`)
                .join("");
              const row = `
            <tr>
              <td><select class="form-control-col item_name"><option value="">Select</option>${options}</select></td>
              <td><input type="number" class="form-control-col rate" min="0" value="0" readonly></td>
              <td><input type="number" class="form-control-col qty" min="1" value="1"></td>
              <td><input type="number" class="form-control-col amount" min="0" value="0" readonly></td>
              <td><input type="text" class="form-control-col remarks" placeholder="Remarks"></td>
              <td><span class="remove-row">✕</span></td>
            </tr>`;
              $("#restaurant_table tbody").append(row);
            };

            $("#add_row_btn").on("click", addRow);
            addRow();

            // remove row
            container.on("click", ".remove-row", function () {
              $(this).closest("tr").remove();
            });

            // auto-calc amount whenever qty changes (rate is fixed)
            container.on("input", ".qty", function () {
              const row = $(this).closest("tr");
              const rate = parseFloat(row.find(".rate").val() || 0);
              const qty = parseFloat(row.find(".qty").val() || 0);
              row.find(".amount").val((rate * qty).toFixed(2));
            });

            // fetch rate from Item Price (price_list_rate)
            container.on("change", ".item_name", function () {
              const item_name = $(this).val();
              const row = $(this).closest("tr");
              const rateInput = row.find(".rate");

              if (item_name) {
                frappe.call({
                  method: "frappe.client.get_value",
                  args: {
                    doctype: "Item Price",
                    filters: { item_code: item_name },
                    fieldname: ["price_list_rate"]
                  },
                  callback: (res) => {
                    const rate = res.message?.price_list_rate || 0;
                    rateInput.val(rate);
                    const qty = parseFloat(row.find(".qty").val() || 1);
                    row.find(".amount").val((rate * qty).toFixed(2));
                  }
                });
              } else {
                rateInput.val("0");
                row.find(".amount").val("0");
              }
            });
          }
        }
      });
    }
    else if (service_type === "Room Service") {
      frappe.call({
        method: "igusto.igusto.page.room_order.room_order.get_service_items",
        callback: (r) => {
          if (r.message) {
            let html = `<label>Room Service Items</label>`;
            r.message.forEach((item, idx) => {
              html += `
          <div class="item-row" data-item-index="${idx}">
            <div>
              <input type="checkbox" id="rs_chk_${idx}" name="service_item" value="${item}">
              <label for="rs_chk_${idx}">${item}</label>
            </div>
            <div class="item-controls">
              <input type="number" id="rs_rate_${idx}" class="form-control-col" readonly style="width:80px;" placeholder="Rate">
            </div>
            <div class="item-controls">
              <input type="number" id="rs_qty_${idx}" class="form-control-col" min="1" value="1" style="width:70px;" disabled>
            </div>
            <div>
              <input type="text" id="rs_remark_${idx}" class="form-control-col" placeholder="Remarks" disabled>
            </div>
          </div>`;
            });
            container.html(html);

            r.message.forEach((item, idx) => {
              // fetch price once when rendering
              frappe.call({
                method: "frappe.client.get_value",
                args: {
                  doctype: "Item Price",
                  filters: { item_code: item },
                  fieldname: "price_list_rate"
                },
                callback: (res) => {
                  if (res.message && res.message.price_list_rate) {
                    $(`#rs_rate_${idx}`).val(res.message.price_list_rate);
                  } else {
                    $(`#rs_rate_${idx}`).val("0");
                  }
                }
              });

              // enable/disable controls
              $(`#rs_chk_${idx}`).on("change", function () {
                const checked = $(this).is(":checked");
                $(`#rs_qty_${idx}, #rs_remark_${idx}`).prop("disabled", !checked);
              });
            });
          }
        }
      });

      // SPA
    } else if (service_type === "Spa") {
      frappe.call({
        method: "igusto.igusto.page.room_order.room_order.get_spa_items",
        callback: (r) => {
          if (r.message) {
            let html = `<label>Spa Services</label>`;
            r.message.forEach((item, idx) => {
              html += `
          <div class="item-row" data-item-index="${idx}">
            <div>
              <input type="checkbox" id="spa_chk_${idx}" name="service_item" value="${item}">
              <label for="spa_chk_${idx}">${item}</label>
            </div>
            <div class="item-controls">
              <input type="number" id="spa_rate_${idx}" class="form-control-col" readonly style="width:80px;" placeholder="Rate">
            </div>
            <div class="item-controls">
              <input type="number" id="spa_qty_${idx}" class="form-control-col" min="1" value="1" style="width:70px;" disabled>
            </div>
            <div>
              <input type="text" id="spa_remark_${idx}" class="form-control-col" placeholder="Remarks" disabled>
            </div>
          </div>`;
            });
            container.html(html);

            r.message.forEach((item, idx) => {
              // fetch price from Item Price
              frappe.call({
                method: "frappe.client.get_value",
                args: {
                  doctype: "Item Price",
                  filters: { item_code: item },
                  fieldname: "price_list_rate"
                },
                callback: (res) => {
                  if (res.message && res.message.price_list_rate) {
                    $(`#spa_rate_${idx}`).val(res.message.price_list_rate);
                  } else {
                    $(`#spa_rate_${idx}`).val("0");
                  }
                }
              });

              $(`#spa_chk_${idx}`).on("change", function () {
                const checked = $(this).is(":checked");
                $(`#spa_qty_${idx}, #spa_remark_${idx}`).prop("disabled", !checked);
              });
            });
          }
        }
      });

      // LAUNDRY
    } else if (service_type === "Laundry") {
      frappe.call({
        method: "igusto.igusto.page.room_order.room_order.get_laundry_items",
        callback: (r) => {
          if (r.message) {
            let html = `<label>Laundry Services</label>`;
            r.message.forEach((item, idx) => {
              html += `
          <div class="item-row" data-item-index="${idx}">
            <div>
              <input type="checkbox" id="laundry_chk_${idx}" name="service_item" value="${item}">
              <label for="laundry_chk_${idx}">${item}</label>
            </div>
            <div class="item-controls">
              <input type="number" id="laundry_rate_${idx}" class="form-control-col" readonly style="width:80px;" placeholder="Rate">
            </div>
            <div class="item-controls">
              <input type="number" id="laundry_qty_${idx}" class="form-control-col" min="1" value="1" style="width:70px;" disabled>
            </div>
            <div>
              <input type="text" id="laundry_remark_${idx}" class="form-control-col" placeholder="Remarks" disabled>
            </div>
          </div>`;
            });
            container.html(html);

            r.message.forEach((item, idx) => {
              frappe.call({
                method: "frappe.client.get_value",
                args: {
                  doctype: "Item Price",
                  filters: { item_code: item },
                  fieldname: "price_list_rate"
                },
                callback: (res) => {
                  $(`#laundry_rate_${idx}`).val(res.message?.price_list_rate || "0");
                }
              });

              $(`#laundry_chk_${idx}`).on("change", function () {
                const checked = $(this).is(":checked");
                $(`#laundry_qty_${idx}, #laundry_remark_${idx}`).prop("disabled", !checked);
              });
            });
          }
        }
      });

      // TRANSPORT
    } else if (service_type === "Transport") {
      frappe.call({
        method: "igusto.igusto.page.room_order.room_order.get_transport_items",
        callback: (r) => {
          if (r.message) {
            let html = `<label>Transport Options</label>`;
            r.message.forEach((item, idx) => {
              html += `
          <div class="item-row" data-item-index="${idx}">
            <div>
              <input type="checkbox" id="trans_chk_${idx}" name="service_item" value="${item}">
              <label for="trans_chk_${idx}">${item}</label>
            </div>
            <div class="item-controls">
              <input type="number" id="trans_rate_${idx}" class="form-control-col" readonly style="width:80px;" placeholder="Rate">
            </div>
            <div class="item-controls">
              <input type="number" id="trans_pass_${idx}" class="form-control-col" min="1" value="1" placeholder="Passengers" style="width:80px;" disabled>
            </div>
            <div>
              <input type="text" id="trans_remark_${idx}" class="form-control-col" placeholder="Remarks" disabled>
            </div>
          </div>`;
            });
            container.html(html);

            r.message.forEach((item, idx) => {
              frappe.call({
                method: "frappe.client.get_value",
                args: {
                  doctype: "Item Price",
                  filters: { item_code: item },
                  fieldname: "price_list_rate"
                },
                callback: (res) => {
                  $(`#trans_rate_${idx}`).val(res.message?.price_list_rate || "0");
                }
              });

              $(`#trans_chk_${idx}`).on("change", function () {
                const checked = $(this).is(":checked");
                $(`#trans_pass_${idx}, #trans_remark_${idx}`).prop("disabled", !checked);
              });
            });
          }
        }
      });


    } else if (service_type === "Other") {
      let html = `
        <div class="form-group">
          <label>Describe Service</label>
          <input type="text" id="service_item_other" class="form-control-col" placeholder="Enter service details">
        </div>
      `;
      container.html(html);
    }
  }

  collect_items(service_type) {
    const items = [];

    if (service_type === "Restaurant") {
      $("#restaurant_table tbody tr").each(function () {
        const item_name = $(this).find(".item_name").val();
        const rate = parseFloat($(this).find(".rate").val() || 0);
        const qty = parseInt($(this).find(".qty").val() || 1);
        const remarks = $(this).find(".remarks").val() || "";

        if (item_name) {
          items.push({ item_name, rate, qty, custom_remarks: remarks });
        }
      });
    }

    else if (["Room Service", "Spa", "Laundry", "Transport"].includes(service_type)) {
      $(`input[name='service_item']`).each(function (idx, el) {
        const $el = $(el);
        if ($el.is(":checked")) {
          const id = $el.attr("id");
          const rate = parseFloat($(`#${id.replace('_chk_', '_rate_')}`).val() || 0);
          const qty = parseInt($(`#${id.replace('_chk_', '_qty_')}`).val() || 1);
          const remark = $(`#${id.replace('_chk_', '_remark_')}`).val() || "";
          const passengers = service_type === "Transport" ? parseInt($(`#${id.replace('_chk_', '_pass_')}`).val() || 1) : null;

          items.push({
            item_name: $el.val(),
            rate,
            quantity: qty,
            passengers,
            custom_remarks: remark
          });
        }
      });
    }

    else if (service_type === "Other") {
      const desc = $("#service_item_other").val();
      if (desc) {
        items.push({
          item_name: desc,
          quantity: parseInt($("#quantity_other").val() || 1),
          custom_remarks: $("#custom_remarks_other").val() || ""
        });
      }
    }

    return items;
  }

  submit_order() {
    const data = {
      room_number: $("#room_number").val(),
      guest: $("#guest_name").val(),
      service_type: $("#service_type").val(),
      delivery_to: $("#delivery_to").val(),
    };

    if (!data.room_number) return frappe.msgprint("Please select Room Number.");
    if (!data.guest) return frappe.msgprint("Guest not found for selected room.");
    if (!data.service_type) return frappe.msgprint("Please select Service Type.");

    const collected = this.collect_items(data.service_type);
    if (!collected.length) return frappe.msgprint("Please select at least one service item.");

    data.items = collected;

    frappe.call({
      method: "igusto.igusto.page.room_order.room_order.create_room_order",
      args: { data: JSON.stringify(data) },
      freeze: true,
      freeze_message: __("Creating Room Order..."),
      callback: (r) => {
        if (r.message) {
          frappe.msgprint({
            title: __("Success"),
            message: `<b>Room Order Created</b>`,
            indicator: "green"
          });
          $("#service_type").val("");
          $("#dynamic_input").empty();
        } else {
          frappe.msgprint("Could not create Room Order. Check server logs.");
        }
      }
    });
  }
}
